lib/cadet/updater/xml_parser.ex

```diff
@@ -0,0 +1,304 @@
+    end
+  rescue
+    e in Timex.Parse.ParseError ->
+      Logger.error("Time does not conform to ISO8601 DateTime: #{e.message}")
+      :error
+
+    # This error is raised by xpath/3 when TASK does not exist (hence is equal to nil)
+    Protocol.UndefinedError ->
+      Logger.error("Missing TASK")
+      :error
+  end
+
+  @spec verify_has_time_offset(%{
+          open_at: DateTime.t() | NaiveDateTime.t(),
+          close_at: DateTime.t() | NaiveDateTime.t()
+        }) :: boolean()
+  defp verify_has_time_offset(%{open_at: open_at, close_at: close_at}) do
+    # Timex.parse!/2 returns NaiveDateTime when offset is not specified, or DateTime otherwise.
+    open_at.__struct__ != NaiveDateTime and close_at.__struct__ != NaiveDateTime
+  end
+
+  @spec process_questions(String.t()) :: {:ok, [map()]} | :error
+  defp process_questions(xml) do
+    default_library = xpath(xml, ~x"//TASK/DEPLOYMENT"e)
+
+    questions_params =
+      xml
+      |> xpath(
+        ~x"//PROBLEMS/PROBLEM"el,
+        type: ~x"./@type"o |> transform_by(&process_charlist/1),
+        max_grade: ~x"./@maxgrade"oi,
+        entity: ~x"."
+      )
+      |> Enum.map(fn param ->
+        with {:no_missing_attr?, true} <-
+               {:no_missing_attr?, not is_nil(param[:type]) and not is_nil(param[:max_grade])},
+             question when is_map(question) <- process_question_by_question_type(param),
+             question when is_map(question) <-
+               process_question_library(question, default_library),
+             question when is_map(question) <- Map.delete(question, :entity) do
+          question
+        else
+          {:no_missing_attr?, false} ->
+            Logger.error("Missing attribute(s) on PROBLEM")
+            :error
+
+          :error ->
+            :error
+        end
+      end)
+
+    if Enum.any?(questions_params, &(&1 == :error)) do
+      :error
+    else
+      {:ok, questions_params}
+    end
+  end
+
+  @spec log_error_bad_changeset(Ecto.Changeset.t(), any()) :: :ok
+  defp log_error_bad_changeset(changeset, entity) do
+    Logger.error("Invalid #{entity} changeset. Error: #{full_error_messages(changeset.errors)}")
+
+    Logger.error("Changeset: #{inspect(changeset, pretty: true)}")
+  end
+
+  @spec process_question_by_question_type(map()) :: map() | :error
+  defp process_question_by_question_type(question) do
+    entity = question[:entity]
+
+    question[:type]
+    |> case do
+      "programming" ->
+        entity
+        |> xpath(
+          ~x"."e,
+          content: ~x"./TEXT/text()" |> transform_by(&process_charlist/1),
+          solution_template: ~x"./SNIPPET/TEMPLATE/text()" |> transform_by(&process_charlist/1),
+          solution: ~x"./SNIPPET/SOLUTION/text()" |> transform_by(&process_charlist/1),
+          autograder:
+            ~x"./SNIPPET/GRADER/text()"l
+            |> transform_by(&Enum.map(&1, fn charlist -> process_charlist(charlist) end))
+        )
+
+      "mcq" ->
+        choices =
```

Comment:
Can we move this and the programming case logic into separate functions? I feel it'll be a lot more readable that way. The newline before `entity` makes it especially confusing for me.